src/olmo_core/nn/rope.py

```diff
@@ -43,32 +43,191 @@ class RoPEType(StrEnum):
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_inv_freqs(theta: int, dim: int, device: torch.device) -> "torch.Tensor":
+        inv_freq = 1.0 / (
+            theta ** (torch.arange(0, dim, 2, device=device, dtype=torch.float) / dim)
+        )
+        return inv_freq
+
+
+@dataclass
+class PIRopeScalingConfig(RoPEScalingConfig):
+    """
+    Position-Interpolation (PI) RoPE scaling from Chen et al. (https://arxiv.org/pdf/2306.15595)
+
+    Interpolate the rotary angles instead of extrapolating them when the context window at
+    inference time exceeds the window used during training. In practice, this amounts to linearly
+    *compressing* the original position indices by a constant factor ``factor``.
+    """
+
+    factor: float = 2.0
+    """Context expansion multiplier. If factor = 1, reduces to vanilla RoPE."""
+
+    def compute_scaled_inv_freq(
+        self, theta: int, dim: int, device: torch.device
+    ) -> tuple["torch.Tensor", float]:
+        inv_freq = self.compute_inv_freqs(theta, dim, device)
+
+        # Positional-interpolation scales the *positions* by 1/factor. This is
+        # equivalent to scaling the inverse frequencies by the same amount.
+        if self.factor != 1.0:
+            inv_freq = inv_freq / self.factor
+
+        return inv_freq, self.attention_rescale_factor
+
+
+@dataclass
+class PerFrequencyRoPEScalingConfig(RoPEScalingConfig):
+    """
+    Per-frequency RoPE scaling (aka "Step-wise" scaling or Llama-3.1 scaling).
+
+    Reference: https://huggingface.co/meta-llama/Llama-3.1-8B/blob/refs%2Fpr%2F3/README.md
+
+    Scales RoPE to longer sequence lengths by interpolating between high- and low-frequency components.
+    1. **High-frequency band** (short wavelengths) – keeps the original frequencies unchanged.
+        These correspond to the very first dimensions of the rotary embedding and already encode
+        short-range ordering well.
+    2. **Low-frequency band** (long wavelengths) – divides the original inverse frequency by
+        ``factor`` (equivalently, multiplies the wavelength by ``factor``).  This has the effect of
+        spreading the very low frequencies across a longer context window (similar to PI scaling).
+    3. **Medium-frequency band** – linearly interpolates (in inverse-frequency space) between the
+        unscaled and the fully-scaled value so that the full spectrum changes smoothly.
     """
 
     factor: float = 32.0
+    """Context expansion multiplier applied to the long-wavelength part of the spectrum."""
+
     low_freq_factor: float = 1.0
+    """
+    Denominator that determines the *low-frequency* wavelength cut-off
+    (a larger value moves the cut-off towards shorter wavelengths).
+    """
+
     high_freq_factor: float = 4.0
+    """
+    Denominator that determines the *high-frequency* wavelength cut-off
+    (a smaller value keeps more of the very short wavelengths untouched).
+    """
+
```

Comment:
it would be nice if these params used the same name / interpretation as the YaRN params that control the same thing... minimally, I'd vote for calling them something other than factor because we also have `factor` above that does a different thing

(I know these names are probably holdovers from the methods, though-- if this is how huggingface does the two methods then maybe it's better to be consistent with them rather than internally consistent)